File: 🟡 Medium/241 - Different Ways to Add Parentheses/_241DifferentWaystoAddParentheses.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {
        vector<int> res;

        for (int i = 0; i < expression.length(); i++) {
            char ch = expression[i];

            if (isdigit(ch)) {
                continue;
            }

            vector<int> left, right;
            left = diffWaysToCompute(expression.substr(0, i));
            right = diffWaysToCompute(expression.substr(i + 1));

            for (auto& j : left) {
                for (auto& k : right) {
                    if (ch == '+') {
                        res.push_back(j + k);
                    } else if (ch == '-') {
                        res.push_back(j - k);
                    } else {
                        res.push_back(j * k);
                    }
                }
            }
        }

        if (res.empty()) {
            res.push_back(stoi(expression));
        }

        return res;
    }
};

File: 🟡 Medium/241 - Different Ways to Add Parentheses/_241DifferentWaystoAddParentheses.js
/**
 * @param {string} expression
 * @return {number[]}
 */
var diffWaysToCompute = function (expression) {
    let res = [];

    for (let i = 0; i < expression.length; i++) {
        let ch = expression[i];
        if (ch == '+' || ch == '-' || ch == '*') {
            let left = diffWaysToCompute(expression.substring(0, i));
            let right = diffWaysToCompute(expression.substring(i + 1));

            for (let j of left) {
                for (let k of right) {
                    if (ch == '+') {
                        res.push(j + k);
                    } else if (ch == '-') {
                        res.push(j - k);
                    } else {
                        res.push(j * k);
                    }
                }
            }
        }
    }

    if (res.length == 0) {
        res.push(parseInt(expression));
    }

    return res;
};